core/src/main/java/feast/core/model/FeatureTable.java

```diff
@@ -0,0 +1,323 @@
+
+  // Entites to associate the features defined in this FeatureTable with
+  @ManyToMany
+  @JoinTable(
+      name = "feature_tables_entities_v2",
+      joinColumns = @JoinColumn(name = "feature_table_id"),
+      inverseJoinColumns = @JoinColumn(name = "entity_v2_id"))
+  private Set<EntityV2> entities;
+
+  // User defined metadata labels serialized as JSON string.
+  @Column(name = "labels", columnDefinition = "text")
+  private String labelsJSON;
+
+  // Max Age of the Features defined in this Feature Table in seconds
+  @Column(name = "max_age", nullable = false)
+  private long maxAgeSecs;
+
+  // Streaming Feature Source used to obtain data for features from a stream
+  @OneToOne(cascade = CascadeType.ALL)
+  @JoinColumn(name = "stream_source_id", nullable = true)
+  private FeatureSource streamSource;
+
+  // Batched Feature Source used to obtain data for features from a batch of data
+  @OneToOne(cascade = CascadeType.ALL)
+  @JoinColumn(name = "batch_source_id", nullable = true)
+  private FeatureSource batchSource;
+
+  // Autoincrementing version no. of this Feature Table.
+  // Autoincrements every update made to the feature set.
+  @Column(name = "revision", nullable = false)
+  private int revision;
+
+  public FeatureTable() {
+    this.revision = 1;
+  };
+
+  /**
+   * Construct Feature from Protobuf spec representation in the given project with entities
+   * registered in entity repository.
+   *
+   * @param projectName the name of the project that the constructed Feature table belongs.
+   * @param spec the Protobuf spec to contruct the Feature from.
+   * @param entityRepo {@link EntityRepository} used to resolve entity names.
+   * @throws IllegalArgumentException if the Protobuf spec provided is invalid.
+   * @return constructed Feature from the given Protobuf spec.
+   */
+  public static FeatureTable fromProto(
+      String projectName, FeatureTableSpec spec, EntityRepository entityRepo) {
+    FeatureTable table = new FeatureTable();
+    table.setName(spec.getName());
+    table.setProject(new Project(projectName));
+
+    Set<FeatureV2> features =
+        spec.getFeaturesList().stream()
+            .map(featureSpec -> FeatureV2.fromProto(table, featureSpec))
+            .collect(Collectors.toSet());
+    table.setFeatures(features);
+
+    Set<EntityV2> entities =
+        FeatureTable.resolveEntities(
+            projectName, spec.getName(), entityRepo, spec.getEntitiesList());
+    table.setEntities(entities);
+
+    String labelsJSON = TypeConversion.convertMapToJsonString(spec.getLabelsMap());
+    table.setLabelsJSON(labelsJSON);
+
+    table.setMaxAgeSecs(spec.getMaxAge().getSeconds());
+
+    // Configure sources only if set
+    if (!spec.getStreamSource().equals(FeatureSourceSpec.getDefaultInstance())) {
+      table.setStreamSource(FeatureSource.fromProto(spec.getStreamSource()));
+    }
+    if (!spec.getBatchSource().equals(FeatureSourceSpec.getDefaultInstance())) {
```

Comment:
You always need a batch source.

Reply:
Updated to make batch source compulsory.